### ISSUE TITLE
Notes from safe_predict development

### ISSUE THREAD

**Issue by alexpghayes** [tidymodels/model-implementation-principles#2]
I'm following the notes from the prediction section fairly closely as I write tests for `modeltests`.

Recommendations I think we should add:

- The `newdata` argument should default to `NULL`
- `predict` should do something graceful when categorical predictors in `newdata` have novel levels not present in the training data. This should be carefully documented.
- I would also add some comments about missing data. Currently the notes specify the number of rows needs to be the same -- I would add that a reasonable default would be to set predictions to `NA` in these cases (although imputation, etc, are also options). Again I think we just want to remind people to document.
- Add a recommendation that `type` default to a character vector of allowed types, so that the function signature looks like

```r
predict_method(object, newdata = NULL, type = c("class", "prob"), ...) {
  type <- match.arg(type)  # key!
}
```

- Predictions should not depend on which observations are present in `newdata`. Mostly this is a `stats::poly()` and `splines::ns()` trap that people can get caught in:

``` r
fit <- lm(mpg ~ disp + hp + splines::ns(drat, 2), mtcars)

hp <- head(predict(fit, mtcars))
ph <- predict(fit, head(mtcars))

all.equal(hp, ph)
#> [1] "Mean relative difference: 0.1161239"

# loading the splines library fixes this

library(splines)
fit <- lm(mpg ~ disp + hp + ns(drat, 2), mtcars)

hp <- head(predict(fit, mtcars))
ph <- predict(fit, head(mtcars))

all.equal(hp, ph)
#> [1] TRUE
```

Created on 2018-09-04 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

There's something similar for `stats::poly()` although the details escape me at the moment.

### Questions

- Say, for logistic regression, I have class A, with predicted prob 0.4 and class B with predicted prob 0.6 -- do people ever report uncertainty in these class probabilities?

[This SO response](https://stackoverflow.com/questions/37842796/how-to-estimate-uncertainties-in-class-probabilities-or-weights-w-for-sgdclass) seems to suggest that GPs for classification do this? I imagine this sort of reporting might be useful for something like a neural net with softmax probabilities where the softmax probabilities aren't really meaningful,  so you might want to look at bootstrapped class probabilities to sanity check.

- What should the factor levels of the `.pred_class` column be? The same factor levels as the factor in input dataset specified as the outcome? Should factor levels that never appear in the training data get dropped? Etc, etc
